cm31_ntt/src/ntt.rs

```diff
@@ -0,0 +1,1289 @@
+    f: &[CF],
+    precomp: &PrecomputedTwiddles,
+) -> Result<Vec<CF>> {
+    let n = f.len();
+    let mut scratch = vec![CF::zero(); NTT_BLOCK_SIZE_FOR_CACHE];
+    debug_assert!(is_power_of(n as u32, 2), "n must be a power of 2");
+    debug_assert!(n >= 8, "n must be at least 8");
+
+    if n <= 64 {
+        // Use the radix-2 NTT for very small sizes
+        let wn = get_root_of_unity(n);
+        return Ok(ntt_radix_2(f, wn));
+    }
+
+    // Case where n is a power of 8
+    if is_power_of(n as u32, 8) {
+        return Ok(ntt_r8_hybrid_p(f, &mut scratch, precomp));
+    }
+
+    // Case where n is of the form 8^k * 2
+    if is_power_of((n / 2) as u32, 8) {
+        return Ok(ntt_r8_s2_hybrid_p(f, &mut scratch, precomp));
+    }
+
+    // Case where n is of the form 8^k * 4
+    if is_power_of((n / 4) as u32, 8) {
+        return Ok(ntt_r8_s4_hybrid_p(f, &mut scratch, precomp));
+    }
+
+    // This shouldn't happen...
+    Err(anyhow!("Could not perform the NTT"))
+}
+
+pub fn precompute_twiddles(n: usize) -> Result<PrecomputedTwiddles> {
+    ensure!(is_power_of(n as u32, 2), "n must be a power of 2");
+
+    // Case where n is a power of 8
+    if is_power_of(n as u32, 8) {
+        let (precomp_small, precomp_full) = precomp_for_ntt_r8_hybrid_p(n);
+        Ok(PrecomputedTwiddles {
+            small: precomp_small,
+            full: precomp_full,
+            stride: vec![],
+        })
+    }
+
+    // Case where n is of the form 8^k * 2
+    else if is_power_of((n / 2) as u32, 8) {
+        let (precomp_small, precomp_full) = precomp_for_ntt_r8_hybrid_p(n / 2);
+        let precomp_s2 = precomp_s2(n);
+        Ok(PrecomputedTwiddles {
+            small: precomp_small,
+            full: precomp_full,
+            stride: precomp_s2,
+        })
+    }
+
+    // Case where n is of the form 8^k * 4
+    else if is_power_of((n / 4) as u32, 8) {
+        let (precomp_small, precomp_full) = precomp_for_ntt_r8_hybrid_p(n / 4);
+        let precomp_s4 = precomp_s4(n);
+        Ok(PrecomputedTwiddles {
+            small: precomp_small,
+            full: precomp_full,
+            stride: precomp_s4,
+        })
+    } 
+
+    else {
+        Err(anyhow!("n must be a power of 2, or of the form 8^k * 2 or 8^k * 4"))
+    }
+}
+
+/// Performs a radix-8 NTT on the polynomial f.
+/// This is unoptimised as it does not use any precomputed twiddles.
+/// It is also slower than ntt_r8_hybrid_p() as it allocates Vecs at each level of recursion, and doing so at
+/// the lower levels (to handle 8^1 to 8^5) incurs significant memory allocation overhead.
+/// @param f The coefficients of the polynomial to be transformed.
+/// @param w The n-th root of unity, where n is the length of f.
+/// @return The transformed polynomial in evaluation form.
+pub fn ntt_r8_vec(f: &[CF], w: CF) -> Vec<CF> {
+    let n = f.len();
+    debug_assert!(n >= 8, "n must be at least 8");
+    debug_assert!(is_power_of(n as u32, 8), "n must be a power of 8");
```

Comment:
```suggestion
    assert!(n >= 8, "n must be at least 8");
    assert!(is_power_of(n as u32, 8), "n must be a power of 8");
```

I'm always in favour of Result though, but I'm not sure how perf critical this is